content/1_docs/2_cookbook/9_setup/0_understanding-oop/recipe.txt

```diff
@@ -0,0 +1,468 @@
+Title: A very brief intro to object oriented programming in PHP
+
+----
+
+Published: 2020-11-09
+
+----
+
+Description: Understanding classes, objects, member functions and common pitfalls
+
+----
+
+Text:
+
+## Intro
+
+The Kirby core is written in PHP, a scripting language that still powers a majority of websites.
+
+In the past, PHP was mostly written using a procedural approach, with functions that act on data structures.
+
+Today, professional PHP programming is based on an OOP (Object Oriented Programming) architecture which has many advantages over the procedural approach. The idea behind OOP is basically to encapsulate data and the methods that act on it, in order to produce modular, extendable and maintainable software.
+
+This recipe will give you just enough background to understand Kirby's architecture and after all, to understand typical errors thrown in your code, like "calling a member function x on null" or why you can use variables like `$kirby` or `$page` in some contexts but not in others.
+
+## Prerequisites
+
+If you want to follow along with this recipe, you need the following:
+
+- A web server (ideally on your (link: docs/cookbook/setup/development-environment text: local machine))
+- A current version of PHP
+- A Kirby installation ((link: docs/guide/quickstart text: Starterkit or Plainkit))
+- A code/text editor of your choice
+
+## What is OOP?
+
+Object Oriented Programming is a programming paradigm based on the concept of objects that have certain properties and methods that allow interacting with these objects. The properties and methods of such an object are described in a class, which acts as the model or blueprint for such objects, whereas an object is an instance of this class, i.e. a specific embodiment of the class.
+
+In a larger software program, you typically have multiple classes whose objects can interact with each other.
+
+### A simple class example
+
+Let's examine a very simple example to get a basic idea. Here we define a class called `Book` with some properties that should apply to all objects of this class, and methods to work with objects of this class.
+
+```php "book.php"
+<?php
+class Book {
```

Comment:
It would be good to have the class set up according to the [PSR-2 standard](https://www.php-fig.org/psr/psr-2/):

```suggestion
class Book
{
```